provider/core/src/union_provider.rs

```diff
@@ -0,0 +1,51 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::iter::*;
+use crate::prelude::*;
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+
+/// A provider that delegates between multiple underlying [`IterableDynProvider`]s,
+/// depending on whether they support the key. A provider supports a key if it
+/// doesn't return [`DataErrorKind::MissingResourceKey`] on `load_payload` or
+/// `supported_options_for_key`. `impl_dyn_provider` implements this behaviour.
+pub struct UnionDataProvider<M: DataMarker>(pub Vec<Box<dyn IterableDynProvider<M>>>);
```

Comment:
Issue: This already exists; it's called [MultiForkByKeyProvider](https://unicode-org.github.io/icu4x-docs/doc/icu_provider/fork/by_key/struct.MultiForkByKeyProvider.html).  We can debate the name (I kind-of like using "union" somewhere in the name), but we should leverage the existing class instead of introducing duplicated functionality. For MultiForkByKeyProvider, we already have `impl<M> DynProvider<M>`, but we don't yet have `impl<M> IterableDynProvider<M>`.

Suggestion: Move your `IterableDynProvider` impl to be on `MultiForkByKeyProvider` instead of `UnionDataProvider` and then replace the usage of `UnionDataProvider` with `MultiForkByKeyProvider`.  Optionally open a follow-up issue to discuss naming.

Note that `MultiForkByKeyProvider` sits on top of `Vec<P>` instead of `Vec<Box<dyn ...>>`. However, the observable behavior should be the same because you can basically set `P = Box<dyn ...>`.

Reply:
I don't think we can set `P = Box<dyn ...>`, as `Box<dyn X>` doesn't implement `X`.